src/databricks/sql/parameters.py

```diff
@@ -0,0 +1,204 @@
+        if self._type is None:
+            self._infer_type()
+        return self._type  # type: ignore
+
+    @type.setter
+    def type(self, value: Type[Enum]) -> None:
+        self._type = value
+
+    def _infer_type(self):
+        value_type = type(self.value)
+        known = TYPE_MAP.get(value_type)
+        if not known:
+            raise NotSupportedError(
+                f"Could not infer parameter type from value: {self.value} - {value_type} \n"
+                "Please specify the type explicitly."
+            )
+        self._type = known
+        self._process_type()
+
+    def _process_type(self):
+        """When self._type requires extra processing, this method performs that processing.
+        Otherwise this method is a no-op.
+        """
+
+        if self._type == DbSqlType.DECIMAL:
+            self._process_decimal()
+        if self._type == DbSqlType.INTEGER:
+            self._process_integer()
+        if self._type == DbSqlType.VOID:
+            self._process_void()
+
+    def _process_decimal(self):
+        """When self._type is DbSqlType.DECIMAL, this method calculates the smallest SQL cast argument that
+        can contain self.value and assigns it to self._type.
+        """
+
+        cast_exp = calculate_decimal_cast_string(self.value)
+        self._type = DbsqlDynamicDecimalType(cast_exp)
+
+    def _process_integer(self):
+        """int() requires special handling becone one Python type can be cast to multiple SQL types (INT, BIGINT, TINYINT)"""
+        self._type = resolve_databricks_sql_integer_type(self.value)
+
+    def _process_void(self):
+        """VOID / NULL types must be passed in a unique way as TSparkParameters with no value"""
+        self.value = None
+
+    def as_tspark_param(self, named: bool) -> TSparkParameter:
+        """Returns a TSparkParameter object that can be passed to the DBR thrift server."""
+        if self.type == DbSqlType.VOID:
+            tsp = TSparkParameter(value=None, type=self.type.value)
+        else:
+            tspark_param_value = TSparkParameterValue(stringValue=str(self.value))
+            tsp = TSparkParameter(type=self.type.value, value=tspark_param_value)
+        if named:
+            tsp.name = self.name
+            tsp.ordinal = False
+        elif not named:
+            tsp.ordinal = True
+        return tsp
+
+    def __str__(self):
+        return f"DbsqlParameter(name={self.name}, value={self.value}, type={self.type.value})"
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __eq__(self, other):
+        return isinstance(other, self.__class__) and self.__dict__ == other.__dict__
+
+
+PrimitiveType = TypeVar(
+    "PrimitiveType",
+    str,
+    int,
+    float,
+    datetime.datetime,
+    datetime.date,
+    bool,
+    decimal.Decimal,
+    Type[None],
+)
+
+ListOfParameters = Union[List[DbsqlParameter], List[PrimitiveType]]
```

Comment:
Minor: prefer ParameterList, ParameterDict.  Prepositions look weird in types :P.